### PR TITLE
Add admin create and delete resource API functions

### LINKED ISSUE #18: Add admin create and delete resource API functions
To allow Octave applications re-architecture without breaking change of the resource tree, we need to be able to create and delete resources in a given namespace.

API functions to create

- `admin_CreateInput`
- `admin_CreateOutput`
- `admin_DeleteResource`

Those API functions will also be required

- `admin_SetJsonExample`
- `admin_MarkOptional`

### PR DESCRIPTION
**Pull request by Nico26-sw**
To allow Octave applications re-architecture without breaking change of the resource tree, we need to be able to create and delete resources in a given namespace.

API functions created

`admin_CreateInput`
`admin_CreateOutput`
`admin_DeleteResource`

Those API functions are also required

`admin_SetJsonExample`
`admin_MarkOptional`

Checklist:
- [x] unit tests added
- [x] formatted following: [edge conventions](https://confluence.sierrawireless.com/display/BROOKLYN/Octave+Edge+Software+Development)
- [x] :bomb: No compatibility break: API or resource format compatible with older blueprints
- [x] :dart: Focuses only on necessary changes

I created Unit tests in Data Hub repo:

```
[fx30]admin(18_admin_create_delete_resources) make tests 
cc -g -m32 -Wall -Werror -o build/test/admintest mock.c main.c ../../components/dataHub/ioService.c ../../components/dataHub/adminService.c ../../components/dataHub/dataHub.c ../../components/dataHub/resource.c ../../components/dataHub/dataSample.c ../../components/dataHub/ioPoint.c ../../components/dataHub/obs.c ../../components/dataHub/resTree.c ../../components/dataHub/handler.c ../../components/dataHub/queryService.c ../../components/dataHub/snapshot.c ../../components/json/json.c ../../components/jsonFormatter/jsonFormatter.c build/test/liblegato/*.o build/test/liblegato/linux/*.o -I. -I../../components/dataHub -I../../components/json -I../../components/jsonFormatter -I/work/repo/legato-r14-fx30/legato/framework/include -I/work/repo/legato-r14-fx30/legato/framework/liblegato/ -I/work/repo/legato-r14-fx30/legato/framework/daemons/linux/ -I/work/repo/legato-r14-fx30/legato/build/wp77xx/framework/include/ -I/work/repo/legato-r14-fx30/legato/build/wp77xx/3rdParty/inc -DUNIT_TEST -lpthread -ldl -lcmocka -lm
build/test/admintest
Apr  6 14:21:52 : -WRN- | _UNKNOWN_[124561]/ T= | timer.c fa_timer_Init() 173 | Using CLOCK_BOOTTIME: alarm wakeups not supported.
[==========] Running 8 test(s).
Apr  6 14:21:52 :  INFO | _UNKNOWN_[124561]/ T=main | dataHub.c initDataHub() 58 | Data Hub started.
[ RUN      ] test_admin_create_delete_input
[       OK ] test_admin_create_delete_input
[ RUN      ] test_admin_create_input_bad_path
Apr  6 14:21:52 : =ERR= | _UNKNOWN_[124561]/ T=main | resTree.c resTree_FindEntryAtAbsolutePath() 421 | Path not absolute.
Apr  6 14:21:52 : =ERR= | _UNKNOWN_[124561]/ T=main | adminService.c ExtractAppNameFromAbsolutePath() 2233 | Path 'app/toto/value' is not absolute.
Apr  6 14:21:52 : =ERR= | _UNKNOWN_[124561]/ T=main | adminService.c admin_CreateInput() 2334 | Failed to extract app name from abs path 'app/toto/value'.
Apr  6 14:21:52 : =ERR= | _UNKNOWN_[124561]/ T=main | adminService.c ExtractAppNameFromAbsolutePath() 2233 | Path '//app/toto/value' is not absolute.
Apr  6 14:21:52 : =ERR= | _UNKNOWN_[124561]/ T=main | adminService.c admin_CreateInput() 2334 | Failed to extract app name from abs path '//app/toto/value'.
Apr  6 14:21:52 : =ERR= | _UNKNOWN_[124561]/ T=main | adminService.c ExtractAppNameFromAbsolutePath() 2233 | Path '/toto/value' is not absolute.
Apr  6 14:21:52 : =ERR= | _UNKNOWN_[124561]/ T=main | adminService.c admin_CreateInput() 2334 | Failed to extract app name from abs path '/toto/value'.
Apr  6 14:21:52 : =ERR= | _UNKNOWN_[124561]/ T=main | resTree.c resTree_FindEntryAtAbsolutePath() 421 | Path not absolute.
Apr  6 14:21:52 : =ERR= | _UNKNOWN_[124561]/ T=main | adminService.c ExtractAppNameFromAbsolutePath() 2233 | Path 'toto/value' is not absolute.
Apr  6 14:21:52 : =ERR= | _UNKNOWN_[124561]/ T=main | adminService.c admin_CreateInput() 2334 | Failed to extract app name from abs path 'toto/value'.
[       OK ] test_admin_create_input_bad_path
[ RUN      ] test_admin_create_input_duplicate
[       OK ] test_admin_create_input_duplicate
[ RUN      ] test_admin_create_delete_output
[       OK ] test_admin_create_delete_output
[ RUN      ] test_admin_create_output_bad_path
Apr  6 14:21:52 : =ERR= | _UNKNOWN_[124561]/ T=main | resTree.c resTree_FindEntryAtAbsolutePath() 421 | Path not absolute.
Apr  6 14:21:52 : =ERR= | _UNKNOWN_[124561]/ T=main | adminService.c ExtractAppNameFromAbsolutePath() 2233 | Path 'app/toto/value' is not absolute.
Apr  6 14:21:52 : =ERR= | _UNKNOWN_[124561]/ T=main | adminService.c admin_CreateOutput() 2421 | Failed to extract app name from abs path 'app/toto/value'.
Apr  6 14:21:52 : =ERR= | _UNKNOWN_[124561]/ T=main | adminService.c ExtractAppNameFromAbsolutePath() 2233 | Path '//app/toto/value' is not absolute.
Apr  6 14:21:52 : =ERR= | _UNKNOWN_[124561]/ T=main | adminService.c admin_CreateOutput() 2421 | Failed to extract app name from abs path '//app/toto/value'.
Apr  6 14:21:52 : =ERR= | _UNKNOWN_[124561]/ T=main | adminService.c ExtractAppNameFromAbsolutePath() 2233 | Path '/toto/value' is not absolute.
Apr  6 14:21:52 : =ERR= | _UNKNOWN_[124561]/ T=main | adminService.c admin_CreateOutput() 2421 | Failed to extract app name from abs path '/toto/value'.
Apr  6 14:21:52 : =ERR= | _UNKNOWN_[124561]/ T=main | resTree.c resTree_FindEntryAtAbsolutePath() 421 | Path not absolute.
Apr  6 14:21:52 : =ERR= | _UNKNOWN_[124561]/ T=main | adminService.c ExtractAppNameFromAbsolutePath() 2233 | Path 'toto/value' is not absolute.
Apr  6 14:21:52 : =ERR= | _UNKNOWN_[124561]/ T=main | adminService.c admin_CreateOutput() 2421 | Failed to extract app name from abs path 'toto/value'.
[       OK ] test_admin_create_output_bad_path
[ RUN      ] test_admin_create_output_duplicate
[       OK ] test_admin_create_output_duplicate
[ RUN      ] test_admin_mark_optional
[       OK ] test_admin_mark_optional
[ RUN      ] test_admin_set_json_example
Apr  6 14:21:52 : -WRN- | _UNKNOWN_[124561]/ T=main | resource.c res_Destruct() 310 | Resource had a JSON example value.
Apr  6 14:21:52 : -WRN- | _UNKNOWN_[124561]/ T=main | resource.c res_Destruct() 310 | Resource had a JSON example value.
Apr  6 14:21:52 : -WRN- | _UNKNOWN_[124561]/ T=main | resource.c res_Destruct() 310 | Resource had a JSON example value.
Apr  6 14:21:52 : -WRN- | _UNKNOWN_[124561]/ T=main | resource.c res_Destruct() 310 | Resource had a JSON example value.
Apr  6 14:21:52 : -WRN- | _UNKNOWN_[124561]/ T=main | resource.c res_Destruct() 310 | Resource had a JSON example value.
[       OK ] test_admin_set_json_example
[==========] 8 test(s) run.
[  PASSED  ] 8 test(s).

```

I also did some tests on target:

```
static void testTriggerHandler(double timestamp, void *contextPtr)
{
    LE_UNUSED(contextPtr);

    admin_PushNumeric("/app/year/may/value", IO_NOW, timestamp);
    admin_PushNumeric("/app/year/june/value", IO_NOW, timestamp);
}

COMPONENT_INIT
{
    // TEST
    admin_CreateInput("/app/toto/value", IO_DATA_TYPE_JSON, "");
    admin_CreateInput("/app/titi/tutu/value", IO_DATA_TYPE_JSON, "");
    admin_CreateInput("/app/tata/tutu/value", IO_DATA_TYPE_JSON, "");

    admin_CreateInput("/app/year/january/value", IO_DATA_TYPE_TRIGGER, "");

    admin_CreateInput("/app/year/february/value", IO_DATA_TYPE_NUMERIC, "");
    admin_CreateInput("/app/year/march/value", IO_DATA_TYPE_JSON, "");

    admin_CreateOutput("/app/year/april/value", IO_DATA_TYPE_TRIGGER, "");
    admin_AddTriggerPushHandler("/app/year/april/value", testTriggerHandler, NULL);

    admin_CreateOutput("/app/year/may/value", IO_DATA_TYPE_NUMERIC, "");
    admin_MarkOptional("/app/year/may/value");

    admin_CreateOutput("/app/year/june/value", IO_DATA_TYPE_JSON, "");
    admin_MarkOptional("/app/year/june/value");
}
```

### CODE REVIEW

**Comment by jvermillard** [comment]
datahub first unit tests :heart_eyes:

File: components/dataHub/adminService.c [hunk 1]
```diff
@@ -2223,3 +2223,305 @@ void admin_EndUpdate
 
     res_EndUpdate();
 }
+
+static le_result_t ExtractAppNameFromAbsolutePath(const char* path, char* appName, size_t appNameSize)
```
Comment (imorrisonca) [inline components/dataHub/adminService.c:2227]: Please add a header comment

File: components/dataHub/adminService.c [hunk 2]
```diff
@@ -2223,3 +2223,305 @@ void admin_EndUpdate
...
+void admin_MarkOptional
+(
+    const char* path
+        ///< [IN] Absolute resource tree path.
+)
+{
+    resTree_EntryRef_t resRef = resTree_FindEntryAtAbsolutePath(path);
+    if (resRef == NULL)
+    {
+        LE_KILL_CLIENT("Attempt to mark non-existent resource optional at '%s'.", path);
```
Comment (imorrisonca) [inline components/dataHub/adminService.c:2517]: Should we be using LE_KILL_CLIENT here?  We removed it in some other places to help with stability.

**Review by imorrisonca — COMMENTED** [review]
Hi @Nico26-sw,

The changes look good.  Just a few minor comments inserted.

Also, I think we need similar functionality in the psensor module.  There are apps which use the psensor functions instead of io_Create*.  Perhaps that can be done under a separate ticket.

Thanks,
Ian

File: components/dataHub/adminService.c (hunk 2 above)
Comment (Nico26-sw) [reply components/dataHub/adminService.c:2517]: Actually this is copy paste from `io_MarkOptional`
I only updated 
`resTree_EntryRef_t resRef = FindResource(path);`
to
`resTree_EntryRef_t resRef = resTree_FindEntryAtAbsolutePath(path);`
I kept the rest of the code as is.

I agree, I will change it to LE_ERROR.

File: components/dataHub/adminService.c (hunk 1 above)
Comment (Nico26-sw) [reply components/dataHub/adminService.c]: Done

**Review by jvermillard — APPROVED** [review]

**Pull request merged**